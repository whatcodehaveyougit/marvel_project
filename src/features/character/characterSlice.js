import { createSlice } from '@reduxjs/toolkit'

const initialState = {
  characterComics: [],
  isLoading: false,
  error: null
}

// Create slice is a function that you pass 1 object
export const characterSlice = createSlice({
  name: 'characters', // This is the namespace that is used to generate the action names
  initialState,
  reducers: {
    setCharacterComics: (state, action) => {
      state.characterComics = action.payload
    },
    setCharacterComicsLoading: (state, action) => {
      state.isLoading = action.payload
    },
    setCharacterComicsError: (state, action) => {
      state.error = action.payload
      state.isLoading = false
    }
  },
})

// Action creators are generated for each case reducer function
export const { setCharacterComics, setCharacterComicsLoading, setCharacterComicsError } = characterSlice.actions

export default characterSlice.reducer