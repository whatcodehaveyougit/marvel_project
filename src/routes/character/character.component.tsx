import './character.styles.scss'
import { useParams } from 'react-router-dom';
import { useState, useEffect, useContext } from 'react'
import { fetchData } from '../../utils/utils'
import {
    Accordion,
    AccordionSummary,
    AccordionDetails,
    Box,
    Card,
    CardContent,
    Grid,
    Typography
} from '@mui/material/';
import { Character, Comic, CharacterAPICall } from '../../types/types'
import { CharactersContext } from '../../contexts/characters.context';

const CharacterPage = () => {

    const { charactersData } = useContext(CharactersContext)

    const [characterComics, setCharacterComics] = useState<Comic[]>()
    const [characterData, setCharacterData] = useState<Character>()
    const [backgroundImageUrl, setBackgroundImageUrl] = useState<string>();

    const { characterid } = useParams();

    const setBackgroundImageUrlFunction = (characterData: Character): void => {
        const backgroundImageUrlConcatenated = characterData['thumbnail']['path'] + '.' + characterData['thumbnail']['extension']
        setBackgroundImageUrl(backgroundImageUrlConcatenated)
    }

    const fetchComicsData = async () => {
        const result = await fetchData<Comic[]>(`/characters/${characterid}/comics`);
        setCharacterComics(result['data']['results']);
    }

    const fetchCharacterData = async () => {
        const dataFetched = await fetchData<CharacterAPICall>(`/characters/${characterid}`);
        const dataFetchedDrilledInto = dataFetched['data']['results'][0];
        if (dataFetchedDrilledInto) {
            setCharacterData(dataFetchedDrilledInto);
            setBackgroundImageUrlFunction(dataFetchedDrilledInto);
        }
    }

    useEffect(() => {
        // If we are given the character data, use that (it will save us having to do another fetch)
        if (charactersData.length > 0) {
            const character = charactersData.find(character => character.id.toString() === characterid)
            if (character !== undefined) {
                setCharacterData(character);
                setBackgroundImageUrlFunction(character)
            }
        } else {
            fetchCharacterData()
                .catch(console.error)
        }
        fetchComicsData()
            .catch(console.error)
    }, [])


    return (
        <>
            <Card className="character-page">
                <div className="thumbnail-image"
                    style={{ backgroundImage: `url(${backgroundImageUrl})` }}
                >
                    <CardContent className="character-information">
                        <Grid className="character-page-headings">
                            <Typography align="center" variant="h3">{characterData ? characterData.name : null}</Typography>
                            <Typography align="center" variant="h5">List of comics for this character:</Typography>
                        </Grid>
                        <Grid>
                            {
                                characterComics && characterComics.map((comic) => (
                                    <Accordion
                                        className="accordion-character-container"
                                        key={comic.id}
                                    >
                                        <AccordionSummary
                                            // expandIcon={ />}
                                            aria-controls="panel1a-content"
                                            id="panel1a-header"
                                        >
                                            <Box sx={{ fontWeight: 'bold' }}>{comic.title}</Box>
                                        </AccordionSummary>
                                        <AccordionDetails>
                                            <Typography>
                                                {comic.description ?
                                                    comic.description :
                                                    'No description for this comic ;('
                                                }
                                            </Typography>
                                        </AccordionDetails>
                                    </Accordion>
                                ))
                            }
                        </Grid>
                    </CardContent>
                </div>
            </Card>

        </>
    )
}

export default CharacterPage;