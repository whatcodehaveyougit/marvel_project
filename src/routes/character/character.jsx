import { useParams } from "react-router-dom";
import { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import Accordion from "../../components/accordion/accordion.component";
import {
  fetchCharacterComicsAsync,
  fetchCharacterDataAsync,
} from "../../store/character/character.action";
import { generateBackgroundImageUrl } from "../../utils/utils";
import "./character.styles.scss";
import {
  setCharacterComics,
  setCharacterData,
} from "../../store/character/characterSlice";
import {
  selectCharacterComics,
  selectIsCharacterComicsLoading,
  selectCharacterData,
} from "../../store/character/character.selector";
import Spinner from "../../components/spinner/spinner.component";

const Character = () => {
  const dispatch = useDispatch();
  const { characterid } = useParams();
  useEffect(() => {
    dispatch(setCharacterComics([]));
    dispatch(setCharacterData(null));
    dispatch(fetchCharacterComicsAsync(characterid));
    dispatch(fetchCharacterDataAsync(characterid));
  }, [dispatch, characterid]);

  const characterComicsArr = useSelector(selectCharacterComics);
  const characterComicsLoading = useSelector(selectIsCharacterComicsLoading);
  const characterData = useSelector(selectCharacterData);

  return (
    <div className="character-page bg-gray-100">
      <div
        className="thumbnail-image bg-cover bg-center"
        style={{
          backgroundImage: `url(${generateBackgroundImageUrl(characterData)})`,
        }}
      >
        <div className="character-information p-8">
          <div className="character-page-headings">
            <h3 className="text-3xl text-center">
              {characterData ? characterData.name : null}
            </h3>
            <h5 className="text-lg text-center">
              List of comics for this character:
            </h5>
          </div>
          <div>
            {characterComicsLoading ? (
              <Spinner />
            ) : (
              characterComicsArr &&
              characterComicsArr.map((comic) => (
                <Accordion
                  title={comic.title}
                  description={comic.description}
                  key={comic.id}
                />
              ))
            )}
          </div>
        </div>
      </div>
    </div>
  );
};

export default Character;
