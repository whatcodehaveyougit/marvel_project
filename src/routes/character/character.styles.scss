.character-page {
    background-color: transparent !important;
    .character-page-headings {
        padding: 20px 0;
    }
    .character-information {
        background-color: black;
        width: 100%;
        opacity: 0.80;
        color: white;
        min-height: 500px;
        padding: 0px;
    }
    .accordion-character-container {
        margin: 0px !important;
        border: 1px solid #c3b6b6;
        color: white;
        background-color: transparent;
    }
}