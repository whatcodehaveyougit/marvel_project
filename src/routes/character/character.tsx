import { useParams } from "react-router-dom";
import { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import Accordion from "../../components/accordion/accordion.tsx";
import { fetchCharacterComicsAsync } from "../../store/characterComics/character.comics.slice.ts";
import { fetchCharacterDataAsync } from "../../store/character/character.slice.ts";
import { generateBackgroundImageUrl } from "../../utils/utils.ts";
import "./character.styles.scss";
import { setCharacterData } from "../../store/character/character.slice.ts";
import {
  selectCharacterComics,
  selectIsCharacterComicsLoading,
  selectCharacterComicsError,
} from "../../store/characterComics/character.comics.selector.ts";
import { selectCharacterData } from "../../store/character/character.selector.ts";
import Spinner from "../../components/spinner/spinner.component.tsx";

const Character = (): JSX.Element => {
  const dispatch = useDispatch();
  const { characterid } = useParams();
  useEffect(() => {
    dispatch(setCharacterData(undefined));
    dispatch(fetchCharacterComicsAsync(characterid));
    dispatch(fetchCharacterDataAsync(characterid));
  }, [dispatch, characterid]);

  const characterComicsArr = useSelector(selectCharacterComics);
  const characterComicsLoading = useSelector(selectIsCharacterComicsLoading);
  const characterComicsError = useSelector(selectCharacterComicsError);

  const characterData = useSelector(selectCharacterData);

  return (
    <div className="character-page bg-gray-100">
      <div
        className="thumbnail-image bg-cover bg-center"
        style={{
          backgroundImage: `url(${generateBackgroundImageUrl(characterData)})`,
        }}
      >
        <div className="character-information p-8">
          <div className="character-page-headings">
            <h3 className="text-3xl text-center">
              {characterData ? characterData.name : null}
            </h3>
            <h5 className="text-lg text-center">
              List of comics for this character:
            </h5>
          </div>
          <div>
            {characterComicsLoading === true && <Spinner />}
            {characterComicsArr &&
              characterComicsLoading !== true &&
              characterComicsArr.map((comic) => (
                <Accordion
                  title={comic.title}
                  description={comic.description}
                  key={comic.id}
                />
              ))}
            {characterComicsError && (
              <p className="text-3xl text-center">{characterComicsError}</p>
            )}
          </div>
        </div>
      </div>
    </div>
  );
};

export default Character;
