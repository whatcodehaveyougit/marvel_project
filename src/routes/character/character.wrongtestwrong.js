import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Character from './character';
import { useSelector } from 'react-redux';

// Mock the react-redux useSelector hook
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
}));

const mockStore = configureStore([]);

const initialState = {
  characters: [
    {
      id: 101133,
      name: 'Hulk',
      description: 'Caught and was transformed into the incredibly powerful creature called the Hulk.',
      thumbnail: {
        path: 'http://i.annihil.us/u/prod/marvel/i/mg/5/a0/538615ca33ab0',
        extension: 'jpg',
      },
    },
  ],
  isLoading: false,
  error: null,
};

describe('Character', () => {
  it('renders correctly', () => {
    // Set up the mock implementation of useSelector to return initialState.characters
    useSelector.mockImplementation((selectorFunction) => selectorFunction(initialState));

    const store = mockStore(initialState);
    const { asFragment } = render(
      <Provider store={store}>
        <Character />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});