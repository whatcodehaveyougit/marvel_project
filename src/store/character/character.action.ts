// import {
//   setCharacterData,
//   setCharacterDataError,
//   setCharacterDataLoading,
// } from "./character.slice.ts";
// import { fetchData } from "../../utils/utils.ts";
// import { store } from "../store.ts";
// import { Dispatch } from "@reduxjs/toolkit";
// import { TCharacter } from "../../types/types.ts";

// const fetchCharacterDataAsync =
//   (characterid: Number) => async (dispatch: Dispatch) => {
//     // Check if characters are already present in the store - performance optimisation

//     const characters = store.getState().characters;
//     if (characters && characters.data?.length === 0) {
//       const apiRouteCharacterData = `/characters/${characterid}`;
//       dispatch(setCharacterDataLoading(true));
//       try {
//         const result = await fetchData(apiRouteCharacterData);
//         const resultData = result[0];
//         dispatch(setCharacterData(resultData));
//         dispatch(setCharacterDataLoading(false));
//       } catch (error) {
//         dispatch(setCharacterDataLoading(false));
//         dispatch(setCharacterDataError(error.message));
//       }
//     } else {
//       const characterData = characters.data?.find(
//         (character: TCharacter) => character.id === Number(characterid)
//       );
//       dispatch(setCharacterData(characterData));
//       dispatch(setCharacterDataLoading(false));
//     }
//   };

// export { fetchCharacterDataAsync };
