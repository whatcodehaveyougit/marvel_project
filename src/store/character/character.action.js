import { setCharacterComics, setCharacterComicsError, setCharacterComicsLoading  } from "./characterSlice";
import { setCharacterData, setCharacterDataError, setCharacterDataLoading } from "./characterSlice";
import { fetchData } from "../../utils/utils";
import { useSelector } from "react-redux";
import { selectCharacters } from "../characters/characters.selector";


//  A thunk is a function which returns a function which takes dispatch
const fetchCharacterComicsAsync = (characterid) => async (dispatch) => {
  dispatch(setCharacterComicsLoading(true));
  const apiRouteComicsData = `/characters/${characterid}/comics`;
  try {
    const result = await fetchData(apiRouteComicsData);
    const resultData = result["data"]["results"];
    dispatch(setCharacterComics(resultData));
    dispatch(setCharacterComicsLoading(false));
  } catch (error) {
    dispatch(setCharacterComicsError(error));
    dispatch(setCharacterComicsLoading(false));
  }
}


// This does not working....
// May need to pass it as an argument to the function
// const charactersData = useSelector(selectCharacters);

// Try and get Character data locally first if it is already loaded in the character array ?
// const character = charactersData.find(
//   (character) => character.id === Number(characterid)
// );

const fetchCharacterDataAsync = (characterid) => async (dispatch) => {
  const apiRouteCharacterData = `/characters/${characterid}`;
  dispatch(setCharacterDataLoading(true));
  try {
    const result = await fetchData(apiRouteCharacterData);
    const resultData = result["data"]["results"][0];
    dispatch(setCharacterData(resultData));
    dispatch(setCharacterDataLoading(false));

  } catch (error) {
    dispatch(setCharacterDataLoading(false));
    dispatch(setCharacterDataError(error));
  }
};


export { fetchCharacterComicsAsync, fetchCharacterDataAsync };