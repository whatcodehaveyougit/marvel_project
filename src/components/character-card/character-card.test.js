
import { render, fireEvent, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import CharacterCard from './character-card.componet';
import character from './character.json';
import character2 from './character2.json';
import { BrowserRouter } from 'react-router-dom';

it('renders correctly', () => {

  const tree = renderer
    .create(<BrowserRouter><CharacterCard character={character} /></BrowserRouter>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly with another character', () => {
  // const tree = renderer
  //   .create(<CharacterCard character={character2} />)
  //   .toJSON();
  // expect(tree).toMatchSnapshot();
});

// it('displays the clicked element', () => {
//   render(<Accordion title="title example" description="example description" />);
//   const accordionSummary = screen.getByTestId('accordion-summary');

//   // Clicking on the accordion summary should display the description
//   fireEvent.click(accordionSummary);
//   const descriptionElement = screen.getByText('example description');
//   expect(descriptionElement).toBeInTheDocument();

//   // Clicking again should hide the description
//   fireEvent.click(accordionSummary);
//   expect(descriptionElement).not.toBeInTheDocument();
// });