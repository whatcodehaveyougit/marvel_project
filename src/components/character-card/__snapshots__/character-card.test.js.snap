// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="character-card relative h-full text-center"
>
  <img
    alt="3-D Man"
    className="w-full h-120 object-cover"
    src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg"
  />
  <div
    className="absolute bottom-0 left-1/2 transform -translate-x-1/2 w-full bg-white"
  >
    <div
      className="p-4"
    >
      <h5
        className="text-xl font-bold"
      >
        3-D Man
      </h5>
      <div
        className="container-1011334 max-lines"
      >
        <div>
          <p
            className="text-gray-500"
          >
            No superpowers ;(
          </p>
        </div>
      </div>
    </div>
    <div
      className="text-center mb-4"
    >
      <a
        className="text-white bg-blue-500 hover:bg-blue-700 font-bold py-2 px-4 rounded block"
        href="/character/1011334"
        onClick={[Function]}
      >
        Find out more
      </a>
    </div>
  </div>
</div>
`;
